### PR TITLE
NMK16 big update

### PR DESCRIPTION
**Pull request by gamez-fan**
Code via iq_132 and reworked for the MAME-2003 core

### CODE REVIEW

**Comment by gamez-fan** [comment]
Full WIP to follow

**Comment by Darknior** [comment]
WOAW incredible, thanks a lot :)

**Comment by Wilstorm** [comment]
Thanks @gamez-fan should these new games be in the nmk16.c driver? I can't seem to find "Air Attack" or "Saboten Bombers (set 2). Several of them have "GAME_NOT_WORKING", "GAME_NO_SOUND" or "GAME_UNEMULATED_PROTECTION" flags. Could you specify if they are the parent of clone that is working? A few have both a parent and clones.

Also is there any chance that the "name" column after the manufacturer could be lined up via TAB? The column is currently set at varying depths and it's hard to follow as nothing lines up.

**Comment by gamez-fan** [comment]
On the commit click view the whole file and that will display the full newly updated driver which
will have all the info you require.

**Comment by Wilstorm** [comment]
@gamez-fan well shoot I appreciate what you're doing but nothing is streamlined or some type of workflow. I have no idea what parent, clones or both have been tested as working even if looking at your commits and you don't seem to want to document it very well and what is documented is completely wrong. Some of these games you listed in this update don't even exist in the driver like Air Attack or Saboten Bombers (set 2).

Other games are not tested at all like Pochi & Nyaa. You said yourself those ROMs haven't been around for 15 years and you don't have them so there's no way you've tested some of these games before committing changes. I appreciate the quantity but prefer quality.

Knowing which sound fixes are code changes or require new/updated samples would have been nice.

Several of these ROMs are **non-MAME** standard and don't exist in MAME 0.193 or the rollback set. They are incredibly hard if not impossible to find which makes these games unplayable to the average users.

Trying to follow your updates when there's no code structure and proper indentation is impossible to follow. It's a mess in some parts.

Your update process seems to fly by the seat of your pants and update whatever you feel like in the moment with a quick and dirty note of "this works now" but you won't clarify those notes that are for sure incorrect.

My plan was to build a DAT for all your changes but I just want to share my thoughts before bowing out here.

Trying to merge a change DAT with the main DAT needs good notes and documentation to be successful. If you think about it. Say you've changed tharrier for example. A few ROMs are different from the base lr-mame2003 DAT. If any of those ROMs are merge ROMs in clones they need an update too.

Another example is sound samples like nibbler. In the base DAT it has samples attached as a clone of Vanguard. The changes you made it no longer needs or never needed samples. But regardless with the new changes you need to remove them from the base DAT or the new samples and old samples will conflict. When using something like ClrMamePro both ROM sets use vanguard.zip as the sample archive but will have different ROMs listed in the DAT.

I guess if people want to run your updates they will have to hunt and peck to find the correct ROMs as I can guarantee some will never see the light of day or be playable due to being so old and impossible to find.

Thanks and have a great weekend.

**Comment by dankcushions** [comment]
i think it would be best if all these changes (from the very first backport) were shifted to a dev branch, then we could gradually bring them in driver by driver, along with corresponding dat file changes.

right now we (retropie) are getting support issues about broken sound in games with previously working samples, because they now use new samples, but you'd have to hunt through commits or source code to find out what, and where to find them. i think slow and steady wins this race :)

**Comment by Wilstorm** [comment]
>I think slow and steady wins this race :)

I agree make it a dev branch or experimental. I would rather introduce changes slowly one driver at a time making sure they are working correctly. Doing it methodical has to help and also make it easier to track down issues when they arise. Right now it feels like the updates are all over. If we take the time to test the new sets we wouldn't have to change them due to the ROM rarity.

I don't know if it's possible but using official MAME ROMs where possible would be highly beneficial for those that keep a standard set vs. one off hacked ROM sets that don't exist anymore. I spent hours sifting through hacked & patched ROMs looking for matching hashes. Then another good amount of hours using torrents to download partial sets hunting for those missing onesy twosy ROMs. Even a little more time running some of those old V2V patches against ROMs. I narrowed it down to the 4 individual ROMs listed above but that's still 4 games that don't work. I can't imagine everyone wanting to go through all that to get some of these games working.

I don't mean it badly but honestly to tell someone to look at my code if you want to see what's changed isn't quite good enough. I learned quite a bit from doing this project but not so well I know exactly what's working properly after an update and I surely can't test it because I can't even find the ROMs.

With that I do have a working DAT from the first update up to this last with 4 new games thrown in from today. Also a list of all games (and ROM names) sorted alphabetically for reference. I can't figure out Air Attack except it's bootleg of Thunder Dragon or something along those lines but Thunder Dragon is in the list already with Air Attack so I added Thunder Dragon. I believe all the sound fixes are code only. 

I added these 4 games today:
* "Hacha Mecha Fighter"		hachamf
* "Task Force Harrier"		tharrier
* "Saboten Bombers"		sabotenb
* "Thuder Dragon"			tdragon

The 4 games I can't find the remaining ROMs for are Metal Slug 4 (Decrypted), Metal Slug 5 (Decrypted), Pochi & Pynaa & Pochi & Pnyaa (Decrypted). Which are mslug4nd, mslug5nd, pnyaa & pnyaan. I do appreciate the updated dkong & dkongjr samples, my favorite part of all the updates! ;)

**Comment by gamez-fan** [comment]
Right putting aside the rom and sample changes which i admit can throw newbie users offtrack
i have tested all my code changes and released everything i've worked on here on the xbox
via an updated MAME72 build and there has been not one complaint about it breaking anything.

Now as i've said before im used to working with bonafida standalone MAME emulators
not cores that are shoehorned into frontends with all manor of extra input, graphical, sound
and video settings which are external to this MAME core, so i can only conclude that if my
code is good in my two other builds and trust me i've tested it fully then the problems users are
reporting must be connected to the frontends which launch this core on the Rpi

A prime example is the Bally Midway or Sega System 16 games which are fine in the original
MAME78 core but broken when using the frontends on the Rpi why is that do ya think..??
i'll leave that question hanging but although im kinda a little pissed off considering the amount
of graft i've put into this feel free to just dump it all and rollback the core to before my first
commit.

**Comment by dankcushions** [comment]
The user issues are the sample issues, nothing else (as far as I know).

No-one is saying it should be rolled back - I'm saying it should be on a separate branch, so we can fully document each change incrementally.

**Comment by arcadez** [comment]
From reading the Rpi forum i see some people are still bitching about roms, par for the course it 
seems in the Rpi scene :) remember i did offer to provide a working parent for every game i added
and a new sample where required then all you had to do was make pack for use with 
MAME2003-plus libretro core.

But since you wanna go at it the hardway via a datfile for roms that will only exist on a few
sites which alotta Rpi users wont know about i better give ya's this heads up, i have recently
changed the rom support for the games i added to the NeoGeo driver in my alt core.

This is good as i have removed the all the decrypted versions and lined up the roms to
match the version of MAME in which they were originally added removing in the process
all the Kawacks roms and to top it all off they now have their full CRC and SHA info plus
detailed WIP from what versions of MAME the code comes from.

Now if you dont piss me off further by keeping on with "gamezfan hasn't tested all his changes"
bullshit when all them have been tried and tested and then ported to other projects on both the
Xbox and Xbox360 with not one single issue reported then maybe i'll actually tell you were to grab
the roms you cant trackdown or even hand you them personally.
